### PR TITLE
[doc] Update secure boot documentation.

### PR DESCRIPTION
**Pull request by jadephilipoom**
Provide a new description of secure boot that's updated to current procedures and mostly self-contained. The goal here is to guide a reader through the process step-by-step with just enough detail to understand what's happening and why. It's heavily based on the previous content of this page but also incorporates a bit of relevant information from e.g. the Identities and Root Keys page to minimize the need for cross-referencing.

### CODE REVIEW

File: doc/security/specs/secure_boot/index.md [hunk 1]
```diff
@@ -176,8 +190,10 @@ manifest format is required to support:
 [ibex-epmp]: https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
 [identities-keys]: {{< relref "/doc/security/specs/identities_and_root_keys" >}}
 [key-manager]: {{< relref "/hw/ip/keymgr/doc" >}}
+[manifest-format]: {{< relref "/sw/device/silicon_creator/rom_ext/docs/manifest" >}}
 [mask-rom-epmp]: {{< relref "/sw/device/silicon_creator/mask_rom/docs/memory_protection" >}}
```
Comment (moidx) [inline doc/security/specs/secure_boot/index.md:194]: This cross-reference is useful along `ibex-epmp`.

**Review by moidx — APPROVED** [review]

File: doc/security/specs/secure_boot/index.md [hunk 2]
```diff
@@ -6,22 +6,96 @@ title: "Secure Boot"
...
+    * The manifest's modulus field identifies which of the stored [Silicon Creator keys][silicon-creator-keys] should be used for signature verification. If there is no matching stored key, or the matching key is not suitable for the device’s lifecycle state, then the boot fails.
+2. Read usage constraints from hardware according to the selector bits from the manifest.
+    * The selector bits can choose all or part of the device ID, and/or information about the state of the device. Because the selected constraints are included in the digest, this means the signer of a `ROM_EXT` image can restrict their signature to only certain devices/states.
+3. Compute the SHA2-256 digest of the selected usage constraints concatenated with the contents of memory between the `ROM_EXT` start and end address given in the manifest.
+4. Check the signature from the manifest against the digest and the selected Silicon Creator public key.
+    * If this step fails, it’s retried just in case of a transient error, but only once. If the second try fails, the boot fails.
```
Comment (alphan) [inline doc/security/specs/secure_boot/index.md:41]: nit: we retry with the second slot, not the same slot.

should this be mentioned somewhere earlier?

File: doc/security/specs/secure_boot/index.md [hunk 3]
```diff
@@ -6,22 +6,96 @@ title: "Secure Boot"
...
+    * Unlock flash execution, configure ePMP so that the `ROM_EXT` region is executable, and jump to the start of `ROM_EXT`.
+
+## `ROM_EXT` {#rom-ext}
+
+The `ROM_EXT` (“ROM extension”) stage is another region of read-only memory that is controlled by the Silicon Creator.
```
Comment (alphan) [inline doc/security/specs/secure_boot/index.md:46]: nit: “ -> "

**Review by alphan — APPROVED** [review]

File: doc/security/specs/secure_boot/index.md (hunk 2 above)
Comment (jadephilipoom) [reply doc/security/specs/secure_boot/index.md:41]: Ah, good point, I'll clarify earlier that there are two `ROM_EXT` slots and that the second try is with the other slot. How does the boot procedure decide which slot to try first? Is it part of the boot policy page or is it chosen somewhere else?

File: doc/security/specs/secure_boot/index.md (hunk 1 above)
Comment (jadephilipoom) [reply doc/security/specs/secure_boot/index.md:194]: Thanks for pointing it out, I'll add that link to the text about ePMP :slightly_smiling_face:

File: doc/security/specs/secure_boot/index.md (hunk 2 above)
Comment (alphan) [reply doc/security/specs/secure_boot/index.md:41]: We pick the one with the newer security version ([here](https://github.com/lowRISC/opentitan/blob/5bb79157b9dea77f485f869938b8ddaa8ec93b59/sw/device/silicon_creator/mask_rom/boot_policy.c#L14)).

File: doc/security/specs/secure_boot/index.md [hunk 4]
```diff
@@ -6,22 +6,97 @@ title: "Secure Boot"
...
+    * In between getting the Silicon Creator identity key and the Silicon Owner intermediate key, check if the boot purpose was "attestation"; if so, generate an attestation certificate for the software/hardware state, sign it with the Silicon Creator identity key, and write it to a special region of memory.
+7. Unlock flash execution, configure ePMP so that the Silicon Creator controlled regions of memory are not writable and the BL0 region is executable, and then jump to the start of BL0.
+
+Once the code has jumped into the Silicon Owner code at BL0, secure boot in its simplest form is complete.
+The Silicon Owner may choose to extend the secure boot process with multiple boot stages of their own; this will differ between device owners, while the stages described here are guaranteed by the Silicon Creator and will be shared by all OpenTitan implementations.
+If any signature verification in the above process fails, or there is any kind of unexpected error, the device will fail to boot.
```
Comment (cdgori) [inline doc/security/specs/secure_boot/index.md:72]: Sorry for this post-merge comment.  Everything here helps a *lot* with clarifying the process, it's nice and clear.

But, should we define what the "fail to boot" behavior/state is more precisely as well?  It seems like it should be consistent for any of the possible failure modes (sig mismatch, or other errors).

We may not know exactly what is yet but when we do, it should probably go in this document.

Comment (jadephilipoom) [reply doc/security/specs/secure_boot/index.md:72]: @alphan , do you have thoughts on this one?

Comment (alphan) [reply doc/security/specs/secure_boot/index.md:72]: Oh, I think I misunderstood @cdgori's comment. I think we have roughly two classes of reasons that can lead to shutdown:
- Any errors (from drivers, e.g. flash_ctrl reports an error, or other code, e.g. failure to find a valid ROM_EXT image in either of the slots)
- Any interrupts or exceptions (we don't expect any in ROM).

Comment (jadephilipoom) [reply doc/security/specs/secure_boot/index.md:72]: > * Any errors (from drivers, e.g. flash_ctrl reports an error, or other code, e.g. failure to find a valid ROM_EXT image in either of the slots)
> * Any interrupts or exceptions (we don't expect any in ROM). For exceptions, please note that we have hardenings that generate exceptions on purpose when we detect something unexpected.

Just to confirm -- both of these things trigger the same boot failure procedure, right?

I think maybe the best way to add this information is to include another paragraph at the top that says something like this:

**Boot failures**. If any step of secure boot cannot be completed (e.g. a driver returns an error code, the ROM_EXT image cannot be found), or there is any kind of interrupt or exception in Mask ROM or ROM_EXT, then the device follows a specific boot-failure process: it triggers an alert, disables critical peripherals such as the key manager and flash, re-scrambles main memory, and resets itself.

WDYT?

Comment (alphan) [reply doc/security/specs/secure_boot/index.md:72]: > Just to confirm -- both of these things trigger the same boot failure procedure, right?

Yes, they all eventually end up in `shutdown_finalize()`. There is a minor difference for interrupts & exceptions that occur very early in the boot process, i.e. before we start executing c code: they are handled by `_asm_exception_handler` which resets the chip and kills the flash. I'm not sure if you want to go into this much detail in this doc. You may want to mention this in parentheses maybe?

 > WDYT?

Your suggestion lgtm. Thanks @jadephilipoom!

**Pull request merged**